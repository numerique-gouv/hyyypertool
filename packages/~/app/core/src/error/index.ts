//

import ModernError from "modern-errors";

//

export class AuthError extends Error {
  constructor(message: string, options?: ErrorOptions) {
    super(message, options);
    this.name = "AuthError";
  }
}

export const BadRequestError = ModernError.subclass("BadRequestError");
export const HTTPError = ModernError.subclass("HTTPError");
export const NotFoundError = ModernError.subclass("NotFoundError");
