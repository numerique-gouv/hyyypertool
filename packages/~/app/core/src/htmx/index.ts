//

import { z } from "zod";

export type Htmx_Header = {
  "HX-Refresh"?: "true" | "false";
  "HX-Location"?: string;
  "HX-Trigger"?: string;
};

// \from https://htmx.org/reference/#events
export const Htmx_Events = z.nativeEnum({
  abort: "htmx:abort", //	send this event to an element to abort a request
  afterOnLoad: "htmx:afterOnLoad", //	triggered after an AJAX request has completed processing a successful response
  afterProcessNode: "htmx:afterProcessNode", //	triggered after htmx has initialized a node
  afterRequest: "htmx:afterRequest", //	triggered after an AJAX request has completed
  afterSettle: "htmx:afterSettle", //	triggered after the DOM has settled
  afterSwap: "htmx:afterSwap", //	triggered after new content has been swapped in
  beforeCleanupElement: "htmx:beforeCleanupElement", //	triggered before htmx disables an element or removes it from the DOM
  beforeHistorySave: "htmx:beforeHistorySave", //	triggered before content is saved to the history cache
  beforeOnLoad: "htmx:beforeOnLoad", //	triggered before any response processing occurs
  beforeProcessNode: "htmx:beforeProcessNode", //	triggered before htmx initializes a node
  beforeRequest: "htmx:beforeRequest", //	triggered before an AJAX request is made
  beforeSend: "htmx:beforeSend", //	triggered just before an ajax request is sent
  beforeSwap: "htmx:beforeSwap", //	triggered before a swap is done, allows you to configure the swap
  configRequest: "htmx:configRequest", //	triggered before the request, allows you to customize parameters, headers
  confirm: "htmx:confirm", //	triggered after a trigger occurs on an element, allows you to cancel (or delay) issuing the AJAX request
  historyCacheError: "htmx:historyCacheError", //	triggered on an error during cache writing
  historyCacheMiss: "htmx:historyCacheMiss", //	triggered on a cache miss in the history subsystem
  historyCacheMissError: "htmx:historyCacheMissError", //	triggered on a unsuccessful remote retrieval
  historyCacheMissLoad: "htmx:historyCacheMissLoad", //	triggered on a successful remote retrieval
  historyRestore: "htmx:historyRestore", //	triggered when htmx handles a history restoration action
  load: "htmx:load", //	triggered when new content is added to the DOM
  noSSESourceError: "htmx:noSSESourceError", //	triggered when an element refers to a SSE event in its trigger, but no parent SSE source has been defined
  onLoadError: "htmx:onLoadError", //	triggered when an exception occurs during the onLoad handling in htmx
  oobAfterSwap: "htmx:oobAfterSwap", //	triggered after an out of band element as been swapped in
  oobBeforeSwap: "htmx:oobBeforeSwap", //	triggered before an out of band element swap is done, allows you to configure the swap
  oobErrorNoTarget: "htmx:oobErrorNoTarget", //	triggered when an out of band element does not have a matching ID in the current DOM
  prompt: "htmx:prompt", //	triggered after a prompt is shown
  pushedIntoHistory: "htmx:pushedIntoHistory", //	triggered after an url is pushed into history
  responseError: "htmx:responseError", //	triggered when an HTTP response error (non-200 or 300 response code) occurs
  sendError: "htmx:sendError", //	triggered when a network error prevents an HTTP request from happening
  sseError: "htmx:sseError", //	triggered when an error occurs with a SSE source
  sseOpen: "htmx:sseOpen", //	triggered when a SSE source is opened
  swapError: "htmx:swapError", //	triggered when an error occurs during the swap phase
  targetError: "htmx:targetError", //	triggered when an invalid target is specified
  timeout: "htmx:timeout", //	triggered when a request timeout occurs
  validation_failed: "htmx:validation:failed", //	triggered when an element fails validation
  validation_halted: "htmx:validation:halted", //	triggered when a request is halted due to validation errors
  validation_validate: "htmx:validation:validate", //	triggered before an element is validated
  xhr_abort: "htmx:xhr:abort", //	triggered when an ajax request aborts
  xhr_loadend: "htmx:xhr:loadend", //	triggered when an ajax request ends
  xhr_loadstart: "htmx:xhr:loadstart", //	triggered when an ajax request starts
  xhr_progress: "htmx:xhr:progress", //	triggered periodically during an ajax request that supports progress events
} as const);

export function hx_trigger_from_body(events: string[]) {
  return events.map((event) => `${event} from:body`);
}
export function hx_include(ids: string[]) {
  return ids.map(prefix_id).join(", ");
}
export function prefix_id(name: string) {
  return `#${name}`;
}

export function is_htmx_request(req: Request) {
  return req.headers.get("hx-request") === "true";
}

export const hx_disabled_form_elements = {
  "hx-disabled-elt": ["button", "input", "textarea"].join(", "),
};
