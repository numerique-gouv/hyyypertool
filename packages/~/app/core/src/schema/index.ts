//

import { z } from "zod";

//

export const Id_Schema = z.string().pipe(z.coerce.number());
export const Entity_Schema = z.object({
  id: Id_Schema,
});

//

export const Pagination_Schema = z.object({
  page: z
    .string()
    .pipe(z.coerce.number().transform((number) => Math.max(1, number)))
    .default("1"),
  page_size: z.string().pipe(z.coerce.number()).default("10"),
});
export type Pagination = z.infer<typeof Pagination_Schema>;

//

export const PAGINATION_ALL_PAGES: Readonly<Pagination> = {
  page: 0,
  page_size: 0,
};

//

export const Search_Schema = z.object({
  q: z.string().default(""),
});
export type Search = z.infer<typeof Search_Schema>;
//

export const DescribedBy_Schema = z.object({
  describedby: z.string(),
});
export type DescribedBy = z.infer<typeof DescribedBy_Schema>;
export const MfaAcrValue_Schema = z.enum([
  "eidas2",
  "eidas3",
  "https://proconnect.gouv.fr/assurance/self-asserted-2fa",
  "https://proconnect.gouv.fr/assurance/consistency-checked-2fa",
]);

export type MfaAcrValue = z.infer<typeof MfaAcrValue_Schema>;
