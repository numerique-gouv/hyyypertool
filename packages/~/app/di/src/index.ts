//

export { asClass, asFunction, asValue, type AwilixContainer } from "awilix";
export { set_injector, set_scope } from "./set_injector";
