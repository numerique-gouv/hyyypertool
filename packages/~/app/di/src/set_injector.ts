//

import { createContainer, type AwilixContainer } from "awilix";
import type { Env } from "hono";
import { createMiddleware } from "hono/factory";

//

export function set_injector<TCradle extends object>(
  fn?: (injector: AwilixContainer<TCradle>) => void,
) {
  type DiEnv = Env & { Variables: { injector: AwilixContainer<TCradle> } };
  return createMiddleware<DiEnv>((c, next) => {
    const container = createContainer<TCradle>({
      strict: true,
    });
    c.set("injector", container);
    if (fn) fn(c.var.injector);
    return next();
  });
}

export function set_scope<TCradle extends object>(
  fn?: (injector: AwilixContainer<TCradle>) => void,
) {
  type DiEnv = Env & { Variables: { injector: AwilixContainer<TCradle> } };
  return createMiddleware<DiEnv>(async (c, next) => {
    if (!c.var.injector)
      throw new Error(`"c.var.injector" is required. use set_injector fist`);
    const parent_injector = c.var.injector;
    const injector = parent_injector.createScope();
    c.set("injector", injector);
    if (fn) fn(injector);
    return next();
  });
}
