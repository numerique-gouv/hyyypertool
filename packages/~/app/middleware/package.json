{
  "name": "@~/app.middleware",
  "version": "1.0.0",
  "private": true,
  "type": "module",
  "exports": {
    ".": {
      "default": "./src/index.ts"
    },
    "./session": {
      "default": "./src/session/index.ts"
    },
    "./vip_list.guard": {
      "default": "./src/vip_list.guard.tsx"
    },
    "./*": {
      "default": "./src/*.ts"
    }
  },
  "scripts": {
    "test": "bun test"
  },
  "dependencies": {
    "@~/app.core": "workspace:*",
    "@~/app.sentry": "workspace:*",
    "@~/app.ui": "workspace:*",
    "@~/app.urls": "workspace:*",
    "@~/moncomptepro.database": "workspace:*",
    "@~/moncomptepro.lib": "workspace:*",
    "hono": "4.6.17",
    "hono-sessions": "0.7.2",
    "openid-client": "5.7.0"
  },
  "devDependencies": {
    "@~/config.typescript": "workspace:*"
  }
}
