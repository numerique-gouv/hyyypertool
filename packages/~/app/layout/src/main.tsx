//

import { z_username } from "@~/app.core/schema/z_username";
import type { UserInfoVariables_Context } from "@~/app.middleware/set_userinfo";
import { urls } from "@~/app.urls";
import type { PropsWithChildren } from "hono/jsx";
import { useRequestContext } from "hono/jsx-renderer";
import { Root_Layout } from "./root";

//
export function Main_Layout({ children }: PropsWithChildren) {
  const {
    var: { userinfo },
  } = useRequestContext<UserInfoVariables_Context>();
  const username = z_username.parse(userinfo);
  return (
    <Root_Layout>
      <div class="flex min-h-full flex-grow flex-col">
        <header role="banner" class="fr-header">
          <div class="fr-header__body">
            <div class="fr-container">
              <div class="fr-header__body-row">
                <Brand />
                <Tools username={username} />
              </div>
            </div>
            {/*  */}
          </div>

          <div class="fr-header__menu fr-modal">
            <div class="fr-container">
              <Nav />
            </div>
          </div>
        </header>
        <div class="relative flex flex-1 flex-col">{children}</div>
      </div>
    </Root_Layout>
  );
}

//

function Brand() {
  return (
    <div class="fr-header__brand fr-enlarge-link">
      <div class="fr-header__brand-top">
        <div class="fr-header__logo">
          <p class="fr-logo">
            République
            <br />
            Française
          </p>
        </div>
      </div>
      <div class="fr-header__service">
        <a href="/" title="Accueil ">
          <p class="fr-header__service-title">Hyyypertool</p>
        </a>
        <p class="fr-header__service-tagline">hyyyyyyyypertool</p>
      </div>
    </div>
  );
}

function Tools({ username }: { username?: string | undefined }) {
  return (
    <div class="fr-header__tools">
      <div class="fr-header__tools-links">
        <ul class="fr-btns-group">
          <li>
            <a
              class="fr-btn fr-btn--sm fr-btn--tertiary-no-outline fr-fi-logout-box-r-line fr-btn--icon-left"
              href={urls.auth.logout.$url().pathname}
            >
              {username}
            </a>
          </li>
        </ul>
      </div>
    </div>
  );
}

function Nav() {
  const { req } = useRequestContext();

  return (
    <nav
      class="fr-nav"
      id="navigation-494"
      role="navigation"
      aria-label="Menu principal"
    >
      <ul class="fr-nav__list">
        <li class="fr-nav__item">
          <a
            aria-current={req.routePath.startsWith("/moderations")}
            class="fr-nav__link"
            href={urls.moderations.$url().pathname}
            target="_self"
          >
            Moderations
          </a>
        </li>
        <li class="fr-nav__item">
          <a
            aria-current={req.routePath.startsWith("/users")}
            class="fr-nav__link"
            href={urls.users.$url().pathname}
            target="_self"
          >
            Utilisateurs
          </a>
        </li>
        <li class="fr-nav__item">
          <a
            aria-current={
              req.routePath.startsWith("/organizations") &&
              !req.routePath.startsWith("/organizations/domains")
            }
            class="fr-nav__link"
            href={urls.organizations.$url().pathname}
            target="_self"
          >
            Organisations
          </a>
        </li>
        <li class="fr-nav__item">
          <a
            aria-current={req.routePath.startsWith("/organizations/domains")}
            class="fr-nav__link"
            href={urls.organizations.domains.$url().pathname}
            target="_self"
          >
            Domaines à verifier
          </a>
        </li>
      </ul>
    </nav>
  );
}
