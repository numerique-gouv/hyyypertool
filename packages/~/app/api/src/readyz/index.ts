//

import env from "@~/app.core/config";
import { identite_pg_database } from "@~/app.middleware/identite_pg";
import { get_zammad_me } from "@~/zammad.lib";
import { to } from "await-to-js";
import { sql } from "drizzle-orm";
import { Hono } from "hono";

//

export default new Hono()
  .get("/", ({ text }) => text(`readyz check passed`))
  .get("/sentry", () => {
    throw new Error("Sentry Check " + new Date().toISOString());
  })
  .get("/zammad", async ({ text }) => {
    const [, user] = await to(get_zammad_me());
    const is_ok = user !== undefined;
    return text(
      [
        `[${is_ok ? "+" : "-"}]zammad identite connection ${user ? "OK" : "FAIL"}`,
        `[${is_ok ? "+" : "-"}]zammad connected as ${user ? user.email : "FAIL"}`,
      ].join("\n"),
    );
  })
  .get(
    "/drizzle/identite",
    identite_pg_database({ connectionString: env.DATABASE_URL }),
    async ({ text, var: { identite_pg } }) => {
      const [, is_ok] = await to(identite_pg.execute(sql`SELECT 1`));
      return text("[+]drizzle identite connection " + (is_ok ? "OK" : "FAIL"));
    },
  );
