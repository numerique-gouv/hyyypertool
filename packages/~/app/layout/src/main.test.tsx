//

import { set_config } from "@~/app.middleware/set_config";
import { set_nonce } from "@~/app.middleware/set_nonce";
import { beforeAll, expect, setSystemTime, test } from "bun:test";
import { Hono } from "hono";
import { jsxRenderer } from "hono/jsx-renderer";
import { set_userinfo } from "../../middleware/src/set_userinfo";
import { Main_Layout } from "./main";

//

beforeAll(() => {
  setSystemTime(new Date("2222-01-01T00:00:00.000Z"));
});

test("Main Layout", async () => {
  const app = new Hono()
    .use(set_config({}))
    .use(set_userinfo({ given_name: "Lara", usual_name: "Croft" }))
    .use(set_nonce("nonce"))
    .use(jsxRenderer(Main_Layout))
    .get("/", (c) => {
      return c.render("âœ…");
    });

  const res = await app.request("/");
  expect(res.status).toBe(200);
  expect(await res.text()).toMatchSnapshot();
});
