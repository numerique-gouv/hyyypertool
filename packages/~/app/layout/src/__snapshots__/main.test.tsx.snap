// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Layout 1`] = `
"<!DOCTYPE html>
    <html
      lang="fr"
      data-fr-scheme="system"
      hx-ext="debug, chunked-transfer, include-vals"
    >
      <head>
        <meta charset="utf-8" />
        <meta
          name="viewport"
          content="width=device-width, initial-scale=1, shrink-to-fit=no"
        />
        <meta name="format-detection" content="telephone=no" />
        <meta name="theme-color" content="#000091" />

        <link
          rel="apple-touch-icon"
          href="/node_modules/@gouvfr/dsfr/dist/favicon/apple-touch-icon.png"
        />
        <link
          rel="icon"
          href="/node_modules/@gouvfr/dsfr/dist/favicon/favicon.svg"
          type="image/svg+xml"
        />
        <link
          rel="shortcut icon"
          href="/node_modules/@gouvfr/dsfr/dist/favicon/favicon.ico"
          type="image/x-icon"
        />
        <link
          rel="manifest"
          href="/node_modules/@gouvfr/dsfr/dist/favicon/manifest.webmanifest"
          crossorigin="use-credentials"
        />

        <!--  -->

        <link
          rel="stylesheet"
          href="/node_modules/animate.css/source/_vars.css"
        />

        <!--  -->

        <link
              rel="stylesheet"
              href="/node_modules/@gouvfr/dsfr/dist/dsfr/dsfr.min.css"
            />

        <!--  -->

        <link
          rel="stylesheet"
          href="/tailwind.css"
        />

        <!--  -->

        <script
          hash="sha256-ZtVO4euNrRnx0KrqoCatLuOIaHv1Z7zi++KgINh8SqM="
          nonce="nonce"
          type="importmap"
        >
          {
            "imports": {
              "@~/app.core/config": "/bundle/config.js",
              "lit": "/node_modules/lit/index.js",
              "lit/": "/node_modules/lit/"
            }
          }
        </script>

        <title>
          Hyyyyypertool
        </title>
      </head>
      <body
        _="
          on every htmx:beforeSend NProgress.start()
          on every htmx:afterOnLoad NProgress.done()
          on every htmx:afterSettle NProgress.done()
        "
        class="flex min-h-screen flex-col"
      >
        <div class="flex flex-1 flex-col"><div class="flex min-h-full flex-grow flex-col"><header role="banner" class="fr-header"><div class="fr-header__body"><div class="fr-container"><div class="fr-header__body-row"><div class="fr-header__brand fr-enlarge-link"><div class="fr-header__brand-top"><div class="fr-header__logo"><p class="fr-logo">RÃ©publique<br/>FranÃ§aise</p></div></div><div class="fr-header__service"><a href="/" title="Accueil "><p class="fr-header__service-title">Hyyypertool</p></a><p class="fr-header__service-tagline">hyyyyyyyypertool</p></div></div><div class="fr-header__tools"><div class="fr-header__tools-links"><ul class="fr-btns-group"><li><a class="fr-btn fr-btn--sm fr-btn--tertiary-no-outline fr-fi-logout-box-r-line fr-btn--icon-left" href="/auth/logout">Lara Croft</a></li></ul></div></div></div></div></div><div class="fr-header__menu fr-modal"><div class="fr-container"><nav class="fr-nav" id="navigation-494" role="navigation" aria-label="Menu principal"><ul class="fr-nav__list"><li class="fr-nav__item"><a class="fr-nav__link" href="/moderations" target="_self">Moderations</a></li><li class="fr-nav__item"><a class="fr-nav__link" href="/users" target="_self">Utilisateurs</a></li><li class="fr-nav__item"><a class="fr-nav__link" href="/organizations" target="_self">Organisations</a></li><li class="fr-nav__item"><a class="fr-nav__link" href="/organizations/domains" target="_self">Domaines Ã  verifier</a></li></ul></nav></div></div></header><div class="relative flex flex-1 flex-col">âœ…</div></div></div>
        <footer class="container mx-auto flex flex-row justify-between p-2">
          <div>Â© 2222 ðŸ‡«ðŸ‡·</div>
          <a
            href="https://github.com/numerique-gouv/hyyypertool/tree/"
            rel="noopener noreferrer"
            target="_blank"
            safe
          >
            <small>version </small>
          </a>
        </footer>
      </body>

      <!--  -->

      <script
        nonce="nonce"
        src="/app/layout/src/_client/nprogress.js"
        type="module"
      ></script>

      <link
        rel="stylesheet"
        href="/node_modules/nprogress/nprogress.css"
      />

      <!--  -->

      <script
            nonce="nonce"
            src="/node_modules/htmx.org/dist/htmx.min.js"
            type="module"
          ></script>

      <meta
        name="htmx-env"
        content="{&quot;defaultSettleDelay&quot;:0,&quot;globalViewTransitions&quot;:true,&quot;historyEnabled&quot;:true,&quot;inlineScriptNonce&quot;:&quot;nonce&quot;}"
      />

      

      <script
        nonce="nonce"
        src="/node_modules/htmx.org/dist/ext/include-vals.js"
        type="module"
      ></script>
      <script
        nonce="nonce"
        src="/node_modules/htmx.org/dist/ext/sse.js"
        type="module"
      ></script>
      <script
        nonce="nonce"
        src="/node_modules/htmx.ext...chunked-transfer/dist/index.js"
        type="module"
      ></script>

      <!--  -->

      <script
            nonce="nonce"
            src="/node_modules/hyperscript.org/dist/_hyperscript.min.js"
          ></script> 
    </html>
  "
`;
